pkg/activator/net/throttler.go

```diff
@@ -699,23 +807,649 @@ func newRevisionThrottler(revID types.NamespacedName,
+		rt.logger.Debugw("All pods removed, capacity set to 0",
+			"revision", rt.revID.String())
+		return
+	}
+
+	// Reset per-pod breakers if needed
+	rt.resetTrackersLocked()
+
+	// Recompute assigned trackers
+	rt.assignedTrackers = rt.recomputeAssignedTrackers(rt.podTrackers)
+
+	// Calculate and update capacity
+	numTrackers := uint64(len(rt.assignedTrackers))
+	activatorCount := rt.numActivators.Load()
+	targetCapacity := rt.calculateCapacity(uint64(backendCount), numTrackers, activatorCount)
+	rt.breaker.UpdateConcurrency(targetCapacity)
+
+	rt.logger.Debugw("Capacity updated",
+		"revision", rt.revID.String(),
+		"backends", backendCount,
+		"assigned", numTrackers,
+		"capacity", targetCapacity)
+}
+
+// resetTrackersLocked resets breaker capacity while holding the lock
+func (rt *revisionThrottler) resetTrackersLocked() {
+	cc := rt.containerConcurrency.Load()
+	if cc == 0 {
+		return
+	}
+
+	for _, t := range rt.podTrackers {
+		if t != nil {
+			t.UpdateConcurrency(cc)
+		}
+	}
+}
+
+// handleExistingPodEvent handles QP events for existing pods
+// Must be called while holding the write lock
+func (rt *revisionThrottler) handleExistingPodEvent(tracker *podTracker, eventType string, qpAuthority bool) {
+	// Always update QP freshness tracking
+	tracker.lastQPUpdate.Store(time.Now().Unix())
+	tracker.lastQPState.Store(eventType)
+
+	// When QP authority is disabled, just log but don't change state
+	if !qpAuthority {
+		rt.logger.Debugw("Received QP event (QP authority disabled, no state change)",
+			"pod-ip", tracker.dest,
+			"event-type", eventType,
+			"current-state", podState(tracker.state.Load()))
+		return
+	}
+
+	oldState := podState(tracker.state.Load())
+	oldRoutable := (oldState == podReady || oldState == podRecovering)
+	stateChanged := false
+
+	// Handle QP event types
+	switch eventType {
+	case "ready":
+		if oldState == podNotReady {
+			if tracker.state.CompareAndSwap(uint32(podNotReady), uint32(podReady)) {
+				stateChanged = true
+				rt.logger.Infow("QP promoted not-ready pod to ready",
+					"pod-ip", tracker.dest)
+			}
+		}
+
+	case "not-ready":
+		if oldState == podReady {
+			if tracker.state.CompareAndSwap(uint32(podReady), uint32(podNotReady)) {
+				stateChanged = true
+				rt.logger.Warnw("QP demoted pod to not-ready (readiness probe failed)",
+					"pod-ip", tracker.dest,
+					"active-requests", tracker.refCount.Load())
+			}
+		}
+
+	case "draining":
+		if tracker.tryDrain() {
+			stateChanged = true
+			rt.logger.Infow("QP initiated pod draining",
+				"pod-ip", tracker.dest,
+				"active-requests", tracker.refCount.Load())
+		} else if oldState == podNotReady {
+			// Pod draining before becoming ready (crash during startup)
+			if tracker.state.CompareAndSwap(uint32(podNotReady), uint32(podDraining)) {
+				tracker.drainingStartTime.Store(time.Now().Unix())
+				stateChanged = true
+				rt.logger.Warnw("Pod draining before ready - crashed during startup",
+					"pod-ip", tracker.dest)
+			}
+		}
```

Comment:
Nit, check the state first. It reads a bit funky to try drain the pod and only then check the status

```suggestion
		if oldState == podNotReady {
			// Pod draining before becoming ready (crash during startup)
			if tracker.state.CompareAndSwap(uint32(podNotReady), uint32(podDraining)) {
				tracker.drainingStartTime.Store(time.Now().Unix())
				stateChanged = true
				rt.logger.Warnw("Pod draining before ready - crashed during startup",
					"pod-ip", tracker.dest)
			}
		} else if tracker.tryDrain() {
			stateChanged = true
			rt.logger.Infow("QP initiated pod draining",
				"pod-ip", tracker.dest,
				"active-requests", tracker.refCount.Load())
		}
```